library/core/src/alloc/global.rs

```diff
@@ -20,24 +20,51 @@ use crate::ptr;
 ///
 /// # Example
 ///
-/// ```no_run
-/// use std::alloc::{GlobalAlloc, Layout, alloc};
-/// use std::ptr::null_mut;
-///
-/// struct MyAllocator;
+/// ```
+/// use std::alloc::{GlobalAlloc, Layout};
+/// use std::cell::UnsafeCell;
+/// use std::process::abort;
+/// use std::sync::atomic::{Ordering, AtomicUsize};
 ///
-/// unsafe impl GlobalAlloc for MyAllocator {
-///     unsafe fn alloc(&self, _layout: Layout) -> *mut u8 { null_mut() }
-///     unsafe fn dealloc(&self, _ptr: *mut u8, _layout: Layout) {}
+/// const ARENA: usize = 100 * 1024;
+/// struct SimpleAllocator {
+///     arena: UnsafeCell<[u8; ARENA]>,
+///     counter: AtomicUsize,
 /// }
 ///
 /// #[global_allocator]
-/// static A: MyAllocator = MyAllocator;
+/// static ALLOCATOR: SimpleAllocator = SimpleAllocator {
+///     arena: UnsafeCell::new([0x55; ARENA]),
+///     counter: AtomicUsize::new(0),
+/// };
 ///
-/// fn main() {
-///     unsafe {
-///         assert!(alloc(Layout::new::<u32>()).is_null())
+/// unsafe impl Sync for SimpleAllocator { }
+///
+/// unsafe impl GlobalAlloc for SimpleAllocator {
+///     unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
```

Comment:
It seems a bit off that the main ~50% of this example is just bit-twiddling that isn't really relevant to understanding the allocator interface.

It might be better to have a non-compiling example that is easier to understand than a compiling one that has a lot of unnecessary complexity.

Reply:
> It seems a bit off that the main ~50% of this example is just bit-twiddling that isn't really relevant to understanding the allocator interface.

The bit twiddling is a shame, yes.  I was going to say "it's a shame we don't have a method on ptr that does this" but I had overlooked `align_offset`.  So I will change it to use that.

> It might be better to have a non-compiling example that is easier to understand than a compiling one that has a lot of unnecessary complexity.

I reread the API specs for `GlobalAlloc` and `Layout` and there are many subtleties which a complete working example gives an opportunity to demonstrate (and discuss in comments).  Prompted by @bluss's comment I've been through the docs again and I think you'll see what I mean when I push an update to this MR...

Reply:
`align_offset` works on pointers and is fiddly to use.  I took a different approach.